src/ferc_xbrl_extractor/instance.py

```diff
@@ -0,0 +1,255 @@
+
+    A period can be instantaneous or a duration of time.
+    """
+
+    instant: bool
+    start_date: Optional[str]
+    end_date: str
+
+    @classmethod
+    def from_xml(cls, elem: Element) -> "Period":
+        """Construct Period from XML element."""
+        instant = elem.find(f"{{{XBRL_INSTANCE}}}instant")
+        if instant is not None:
+            return cls(instant=True, end_date=instant.text)
+
+        return cls(
+            instant=False,
+            start_date=elem.find(f"{{{XBRL_INSTANCE}}}startDate").text,
+            end_date=elem.find(f"{{{XBRL_INSTANCE}}}endDate").text,
+        )
+
+
+class DimensionType(Enum):
+    """
+    Indicate dimension type.
+
+    XBRL contains explicit (all allowable values defined in taxonomy) and typed
+    (value defined in filing) dimensions.
+    """
+
+    EXPLICIT = auto()
+    TYPED = auto()
+
+
+class Axis(BaseModel):
+    """
+    Pydantic model that defines an XBRL Axis.
+
+    Axes (or dimensions, terms are interchangeable in XBRL) are used for identifying
+    individual facts when the entity id, and period are insufficient. All axes will
+    be turned into columns, and be a part of the primary key for the table they
+    belong to.
+    """
+
+    name: str
+    value: str
+    dimension_type: DimensionType
+
+    @validator("name", pre=True)
+    def strip_prefix(cls, name):  # noqa: N805
+        """Strip XML prefix from name."""
+        return name.split(":")[1] if ":" in name else name
+
+    @classmethod
+    def from_xml(cls, elem: Element) -> "Axis":
+        """Construct Axis from XML element."""
+        if elem.tag.endswith("explicitMember"):
+            return cls(
+                name=elem.attrib["dimension"],
+                value=elem.text,
+                dimension_type=DimensionType.EXPLICIT,
+            )
+
+        elif elem.tag.endswith("typedMember"):
+            dim = elem.getchildren()[0]
+            return cls(
+                name=elem.attrib["dimension"],
+                value=dim.text,
+                dimension_type=DimensionType.TYPED,
+            )
+
+
+class Entity(BaseModel):
+    """
+    Pydantic model that defines an XBRL Entity.
+
+    Entities are used to identify individual XBRL facts. An Entity should
+    contain a unique identifier, as well as any dimensions defined for a
+    table.
+    """
+
+    identifier: str
+    dimensions: List[Axis]
+
+    @classmethod
+    def from_xml(cls, elem: Element) -> "Entity":
+        """Construct Entity from XML element."""
+        # Segment node contains dimensions prefixed with xbrldi
+        segment = elem.find(f"{{{XBRL_INSTANCE}}}segment")
+        dims = segment.findall("*") if segment is not None else []
+
+        return cls(
+            identifier=elem.find(f"{{{XBRL_INSTANCE}}}identifier").text,
+            dimensions=[Axis.from_xml(child) for child in dims],
+        )
+
+    def check_dimensions(self, axes: List[str]) -> bool:
```

Comment:
Should these `check_` methods be pydantic validators?

Reply:
These are taking an outside input from a table that is being constructed to see if the context is "valid" for that table

Reply:
Got it thanks.